website/mailing_list/utils.py

```diff
@@ -0,0 +1,425 @@
+
+###############################################################################
+# Decorator
+###############################################################################
+
+def require_mailgun(func):
+    """ Execute MailGun API calls iff API key is set """
+    def wrapper(*args, **kwargs):
+        if settings.MAILGUN_API_KEY:
+            return func(*args, **kwargs)
+        return None
+    return wrapper
+
+###############################################################################
+# List Management Functions
+###############################################################################
+
+@require_mailgun
+def get_info(node_id):
+    """ Returns information about the mailing list from Mailgun
+    :param str node_id: ID of the node in question
+    :returns dict info: mailing list info
+    """
+    resp = requests.get(
+        '{}/{}'.format(MAILGUN_BASE_LISTS_URL, address(node_id)),
+        auth=('api', settings.MAILGUN_API_KEY),
+    )
+    if resp.status_code == 200:
+        return json.loads(resp.text)
+    elif resp.status_code == 404:
+        return None
+    raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+@require_mailgun
+def get_members(node_id):
+    """ Returns member list for mailing list from Mailgun
+    :param str node_id: ID of the node in question
+    :returns dict members: list of members
+    """
+    resp = requests.get(
+        '{}/{}/members'.format(MAILGUN_BASE_LISTS_URL, address(node_id)),
+        auth=('api', settings.MAILGUN_API_KEY),
+    )
+    if resp.status_code == 200:
+        return json.loads(resp.text)
+    raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+@require_mailgun
+def get_list(node_id):
+    """ Returns information about the mailing list from Mailgun
+    :param str node_id: ID of the node in question
+    :returns: info, members: Two dictionaries about list and members
+    """
+    info = get_info(node_id)
+
+    if not info:
+        return None, None
+
+    members = get_members(node_id)
+
+    return info, members
+
+@require_mailgun
+def create_list(node_id):
+    """ Creates a new mailing list on Mailgun with all emails and subscriptions
+    :param str node_id: ID of the node in question
+    """
+    from website.models import Node  # avoid circular import
+    node = Node.load(node_id)
+    resp = requests.post(
+        MAILGUN_BASE_LISTS_URL,
+        auth=('api', settings.MAILGUN_API_KEY),
+        data={
+            'address': address(node_id),
+            'name': list_title(node),
+            'access_level': 'members'
+        }
+    )
+    if resp.status_code != 200:
+        raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+    members_list = []
+    members_list = jsonify_users_list(node.contributors, unsubs=get_unsubscribes(node))
+    members_list[-1].append({'address': address('mailing_list_robot'), 'subscribed': True})  # Routing robot
```

Comment:
I still think paginating the results before you've fully constructed the list of addresses is asking for trouble.

This seems like a case of a method trying to be too clever about too specific a use case. But to meet the current QA deadline, we should put a pin in this and focus our energies on the celery tasks to keep data internally consistent.